modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -2595,13 +2728,127 @@ void GDScriptAnalyzer::reduce_assignment(GDScriptParser::AssignmentNode *p_assig
+			binary_op->variant_op = Variant::OP_BIT_AND;
+		} break;
+		case GDScriptParser::AssignmentNode::OP_BIT_OR: {
+			binary_op->operation = GDScriptParser::BinaryOpNode::OP_BIT_OR;
+			binary_op->variant_op = Variant::OP_BIT_OR;
+		} break;
+		case GDScriptParser::AssignmentNode::OP_BIT_XOR: {
+			binary_op->operation = GDScriptParser::BinaryOpNode::OP_BIT_XOR;
+			binary_op->variant_op = Variant::OP_BIT_XOR;
+		} break;
+	}
+	binary_op->start_line = p_assignment->start_line;
+	binary_op->end_line = p_assignment->end_line;
+	binary_op->start_column = p_assignment->start_column;
+	binary_op->end_column = p_assignment->end_column;
+	binary_op->leftmost_column = p_assignment->leftmost_column;
+	binary_op->rightmost_column = p_assignment->rightmost_column;
+	binary_op->left_operand = p_assignment->assignee;
+	binary_op->right_operand = p_assignment->assigned_value;
+	reduce_binary_op(binary_op);
+	return false;
+}
+
+bool GDScriptAnalyzer::assignable_nullity_guard(GDScriptParser::ExpressionNode *p_assignee, GDScriptParser::ExpressionNode *p_assigned_value, GDScriptParser::Node *p_origin) {
+	if (!p_assignee || !p_assigned_value || !p_origin) {
+		return true;
+	}
+
+	if (p_assignee->type != GDScriptParser::Node::IDENTIFIER) {
+		return true;
+	}
+
+	GDScriptParser::IdentifierNode *identifier = static_cast<GDScriptParser::IdentifierNode *>(p_assignee);
+	// Assigning null or variant to a narrowed identifier should broaden it for safety
+	if (identifier->get_datatype().is_nullable && is_effectively_non_nullable(identifier)) {
+		GDScriptParser::DataType data_type = p_assigned_value->get_datatype();
+		if (is_potentially_null_value(p_assigned_value)) {
+			pop_narrowed_nullables({ identifier });
+		}
+	}
+
+	// Forbid nullables being assigned to non-nullables, unless the non-nullable is an object
+	if (identifier->get_datatype().builtin_type != Variant::OBJECT && !identifier->get_datatype().is_nullable && p_assigned_value->get_datatype().is_nullable && is_effectively_nullable(p_assigned_value)) {
+		// Single out calls to have a more accurate error message
+		if (p_assigned_value->type == GDScriptParser::Node::CALL) {
+			push_nullable_error(R"*(Potentially unsafe to assign the nullable return of "%s()" to non-nullable "%s".)*", {}, p_origin, static_cast<GDScriptParser::CallNode *>(p_assigned_value)->function_name, identifier->name);
+		} else {
+			push_nullable_error(R"*(Potentially unsafe to assign nullable "%s" to non-nullable "%s".)*", {}, p_origin, print_expression(p_assigned_value), identifier->name);
+		}
+		return false;
+	}
+
+	// Special case for container types, forbid assigning a nullable typed array to a non-nullable one (Like assigning an "Array[int?]" to an "Array[int]")
+	if (p_origin->get_datatype().has_container_element_type() && p_assigned_value->get_datatype().has_container_element_type()) {
+		GDScriptParser::DataType identifier_data_type = p_origin->get_datatype().get_container_element_type();
```

Comment:
is this the correct source for the datatype, `p_origin->get_datatype()`? this block only seems to catch:
```gdscript
var a:Array[int?] = [1]
var b:Array[int] = a # analyzer error
```
and not the statement:
```gdscript
b = a # runtime error, no analyzer error
```

I changed all the `p_origin->get_datatype()` references to `identifier->get_datatype()` in this block and it seems ok.
![image](https://github.com/godotengine/godot/assets/86566939/e590f99f-4ef7-43f5-b6a7-9cc20a56095c)